Issue1497 unified cypress folders with other test folders

miguelvaara opened the pull request:
## Reasons for creating this PR
It is clearer for developers if all tests are located in the same directory.

## Link to relevant issue(s), if any
#1497 

- Closes #1497 

## Description of the changes in this PR
Moved the Cypress folder into the same directory as other tests (tests). All template tests were moved under a new directory called template-tests, while the e2e directory was left for e2e tests. Additionally, accessibility tests now have their own directory at the same level as other test directories. Configuration changes were made because changing the directory structure affected the functionality of Cypress Axe

## Known problems or uncertainties in this PR
Merging with the base branch is going to be challenging, as other developers have likely made changes and additions to the existing structure

## Checklist

- [ ] phpUnit tests pass locally with my changes
- [ ] I have added tests that show that the new code works, or tests are not relevant for this PR (e.g. only HTML/CSS changes)
- [x] The PR doesn't reduce accessibility of the front-end code (e.g. tab focus, scaling to different resolutions, use of `.sr-only` class, color contrast)
- [x] The PR doesn't introduce unintended code changes (e.g. empty lines or useless reindentation)

##
- Run `git fetch` 
- `git checkout issue1497-unified-cypress-folders-with-other-test-folders`
- Run `npm install` to ensure you have the latest versions of required packages
- Review the folder structure under the _tests/cypress_ directory
- Start by running `npm run test-accessibility-gui-all` or `npx cypress open` and manually go through all tests in the user interface
- Run all command-line tests together using the command: `npm run test-accessibility-cli-all`
- Run individual command-line tests based on their setup in the package.json file

osma reviewed (changes requested):
I tested this, mainly just by running `npx cypress run`. A few suggestions:

1. Based on the description, I expected to see a directory called `e2e`, but it was removed entirely in this PR. I think this is because Git doesn't track directories separately from files. If you remove all files in a directory, the directory itself will disappear too. So there should be a README or some other placeholder file in the `e2e` directory if we want to keep it.
2. While we are renaming things, I think we should aim for consistency. Now we have directories called `accessibility-tests`, `template-tests` and `e2e`. How about dropping the somewhat redundant `-tests` suffix and simply naming these `accessibility`, `template` and `e2e`?
3. SonarCloud [complains](https://sonarcloud.io/project/issues?resolved=false&types=BUG&pullRequest=1501&id=NatLibFi_Skosmos&open=AYpGTrnauF5ovE0xlyS6) about a mismatch in the number of parameters given to the `checkA11y` function in support/accessibility.js. However, this appears not to be a regression, because the file was simply renamed in this PR without changing the contents.
4. The Cypress test `error.cy.js` is failing. This is also not a regression in this PR, since it is failing on the `skosmos-3` branch as well.
5. There is, however, a regression w.r.t. the accessibility tests. On the `skosmos-3` branch, this is the result of running `npx cypress run` (only error.cy.js is failing):

```
       Spec                                              Tests  Passing  Failing  Pending  Skipped  
  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ ✔  about.cy.js                              00:01        1        1        -        -        - │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  concept.cy.js                             76ms        1        1        -        -        - │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✖  error.cy.js                              00:05        1        -        1        -        - │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  feedback.cy.js                            78ms        1        1        -        -        - │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  global-search.cy.js                       72ms        1        1        -        -        - │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  landing.cy.js                            00:01        1        1        -        -        - │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  vocab-home.cy.js                         00:01        1        1        -        -        - │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  vocab-search.cy.js                        72ms        1        1        -        -        - │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  accessibility-tests/axe-about.cy.js       27ms        1        -        -        1        - │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  accessibility-tests/axe-concept.cy.       23ms        1        -        -        1        - │
  │    js                                                                                          │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  accessibility-tests/axe-feedback.cy       23ms        1        -        -        1        - │
  │    .js                                                                                         │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  accessibility-tests/axe-global-sear       28ms        1        -        -        1        - │
  │    ch.cy.js                                                                                    │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  accessibility-tests/axe-landing.cy.       32ms        1        -        -        1        - │
  │    js                                                                                          │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  accessibility-tests/axe-vocab-home.       30ms        1        -        -        1        - │
  │    cy.js                                                                                       │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  accessibility-tests/axe-vocab-searc       27ms        1        -        -        1        - │
  │    h.cy.js                                                                                     │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  accessibility-tests/axe-vocab.cy.js       25ms        1        -        -        1        - │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘
    ✖  1 of 16 failed (6%)                      00:09       16        7        1        8        -  
```

while on this PR branch, this happens:

```
       Spec                                              Tests  Passing  Failing  Pending  Skipped  
  ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
  │ ✖  accessibility-tests/axe-about.cy.js      868ms        1        -        1        -        - │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  accessibility-tests/axe-concept.cy.       19ms        1        -        -        1        - │
  │    js                                                                                          │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✖  accessibility-tests/axe-feedback.cy      925ms        1        -        1        -        - │
  │    .js                                                                                         │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  accessibility-tests/axe-global-sear       21ms        1        -        -        1        - │
  │    ch.cy.js                                                                                    │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✖  accessibility-tests/axe-landing.cy.      00:01        1        -        1        -        - │
  │    js                                                                                          │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✖  accessibility-tests/axe-vocab-home.      00:01        1        -        1        -        - │
  │    cy.js                                                                                       │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  accessibility-tests/axe-vocab-searc       50ms        1        -        -        1        - │
  │    h.cy.js                                                                                     │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✖  accessibility-tests/axe-vocab.cy.js      00:02        1        -        1        -        - │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  template-tests/about.cy.js               572ms        1        1        -        -        - │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  template-tests/concept.cy.js              75ms        1        1        -        -        - │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✖  template-tests/error.cy.js               00:04        1        -        1        -        - │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  template-tests/feedback.cy.js             73ms        1        1        -        -        - │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  template-tests/global-search.cy.js        57ms        1        1        -        -        - │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  template-tests/landing.cy.js             00:01        1        1        -        -        - │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  template-tests/vocab-home.cy.js          00:01        1        1        -        -        - │
  ├────────────────────────────────────────────────────────────────────────────────────────────────┤
  │ ✔  template-tests/vocab-search.cy.js         73ms        1        1        -        -        - │
  └────────────────────────────────────────────────────────────────────────────────────────────────┘
    ✖  6 of 16 failed (38%)                     00:15       16        7        6        3        -  
```

so five of the accessibility tests are now failing. I think this regression should be fixed before this can be merged.

miguelvaara commented:
> I tested this, mainly just by running `npx cypress run`. A few suggestions:
> 
>     1. Based on the description, I expected to see a directory called `e2e`, but it was removed entirely in this PR. I think this is because Git doesn't track directories separately from files. If you remove all files in a directory, the directory itself will disappear too. So there should be a README or some other placeholder file in the `e2e` directory if we want to keep it.
> 
>     2. While we are renaming things, I think we should aim for consistency. Now we have directories called `accessibility-tests`, `template-tests` and `e2e`. How about dropping the somewhat redundant `-tests` suffix and simply naming these `accessibility`, `template` and `e2e`?
> 
>     3. SonarCloud [complains](https://sonarcloud.io/project/issues?resolved=false&types=BUG&pullRequest=1501&id=NatLibFi_Skosmos&open=AYpGTrnauF5ovE0xlyS6) about a mismatch in the number of parameters given to the `checkA11y` function in support/accessibility.js. However, this appears not to be a regression, because the file was simply renamed in this PR without changing the contents.
> 
>     4. The Cypress test `error.cy.js` is failing. This is also not a regression in this PR, since it is failing on the `skosmos-3` branch as well.
> 
>     5. There is, however, a regression w.r.t. the accessibility tests. On the `skosmos-3` branch, this is the result of running `npx cypress run` (only error.cy.js is failing):
> 
> 
> ```
>        Spec                                              Tests  Passing  Failing  Pending  Skipped  
>   ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
>   │ ✔  about.cy.js                              00:01        1        1        -        -        - │
>   ├────────────────────────────────────────────────────────────────────────────────────────────────┤
>   │ ✔  concept.cy.js                             76ms        1        1        -        -        - │
>   ├────────────────────────────────────────────────────────────────────────────────────────────────┤
>   │ ✖  error.cy.js                              00:05        1        -        1        -        - │
>   ├────────────────────────────────────────────────────────────────────────────────────────────────┤
>   │ ✔  feedback.cy.js                            78ms        1        1        -        -        - │
>   ├────────────────────────────────────────────────────────────────────────────────────────────────┤
>   │ ✔  global-search.cy.js                       72ms        1        1        -        -        - │
>   ├────────────────────────────────────────────────────────────────────────────────────────────────┤
>   │ ✔  landing.cy.js                            00:01        1        1        -        -        - │
>   ├────────────────────────────────────────────────────────────────────────────────────────────────┤
>   │ ✔  vocab-home.cy.js                         00:01        1        1        -        -        - │
>   ├────────────────────────────────────────────────────────────────────────────────────────────────┤
>   │ ✔  vocab-search.cy.js                        72ms        1        1        -        -        - │
>   ├────────────────────────────────────────────────────────────────────────────────────────────────┤
>   │ ✔  accessibility-tests/axe-about.cy.js       27ms        1        -        -        1        - │
>   ├────────────────────────────────────────────────────────────────────────────────────────────────┤
>   │ ✔  accessibility-tests/axe-concept.cy.       23ms        1        -        -        1        - │
>   │    js                                                                                          │
>   ├────────────────────────────────────────────────────────────────────────────────────────────────┤
>   │ ✔  accessibility-tests/axe-feedback.cy       23ms        1        -        -        1        - │
>   │    .js                                                                                         │
>   ├────────────────────────────────────────────────────────────────────────────────────────────────┤
>   │ ✔  accessibility-tests/axe-global-sear       28ms        1        -        -        1        - │
>   │    ch.cy.js                                                                                    │
>   ├────────────────────────────────────────────────────────────────────────────────────────────────┤
>   │ ✔  accessibility-tests/axe-landing.cy.       32ms        1        -        -        1        - │
>   │    js                                                                                          │
>   ├────────────────────────────────────────────────────────────────────────────────────────────────┤
>   │ ✔  accessibility-tests/axe-vocab-home.       30ms        1        -        -        1        - │
>   │    cy.js                                                                                       │
>   ├────────────────────────────────────────────────────────────────────────────────────────────────┤
>   │ ✔  accessibility-tests/axe-vocab-searc       27ms        1        -        -        1        - │
>   │    h.cy.js                                                                                     │
>   ├────────────────────────────────────────────────────────────────────────────────────────────────┤
>   │ ✔  accessibility-tests/axe-vocab.cy.js       25ms        1        -        -        1        - │
>   └────────────────────────────────────────────────────────────────────────────────────────────────┘
>     ✖  1 of 16 failed (6%)                      00:09       16        7        1        8        -  
> ```
> 
> while on this PR branch, this happens:
> 
> ```
>        Spec                                              Tests  Passing  Failing  Pending  Skipped  
>   ┌────────────────────────────────────────────────────────────────────────────────────────────────┐
>   │ ✖  accessibility-tests/axe-about.cy.js      868ms        1        -        1        -        - │
>   ├────────────────────────────────────────────────────────────────────────────────────────────────┤
>   │ ✔  accessibility-tests/axe-concept.cy.       19ms        1        -        -        1        - │
>   │    js                                                                                          │
>   ├────────────────────────────────────────────────────────────────────────────────────────────────┤
>   │ ✖  accessibility-tests/axe-feedback.cy      925ms        1        -        1        -        - │
>   │    .js                                                                                         │
>   ├────────────────────────────────────────────────────────────────────────────────────────────────┤
>   │ ✔  accessibility-tests/axe-global-sear       21ms        1        -        -        1        - │
>   │    ch.cy.js                                                                                    │
>   ├────────────────────────────────────────────────────────────────────────────────────────────────┤
>   │ ✖  accessibility-tests/axe-landing.cy.      00:01        1        -        1        -        - │
>   │    js                                                                                          │
>   ├────────────────────────────────────────────────────────────────────────────────────────────────┤
>   │ ✖  accessibility-tests/axe-vocab-home.      00:01        1        -        1        -        - │
>   │    cy.js                                                                                       │
>   ├────────────────────────────────────────────────────────────────────────────────────────────────┤
>   │ ✔  accessibility-tests/axe-vocab-searc       50ms        1        -        -        1        - │
>   │    h.cy.js                                                                                     │
>   ├────────────────────────────────────────────────────────────────────────────────────────────────┤
>   │ ✖  accessibility-tests/axe-vocab.cy.js      00:02        1        -        1        -        - │
>   ├────────────────────────────────────────────────────────────────────────────────────────────────┤
>   │ ✔  template-tests/about.cy.js               572ms        1        1        -        -        - │
>   ├────────────────────────────────────────────────────────────────────────────────────────────────┤
>   │ ✔  template-tests/concept.cy.js              75ms        1        1        -        -        - │
>   ├────────────────────────────────────────────────────────────────────────────────────────────────┤
>   │ ✖  template-tests/error.cy.js               00:04        1        -        1        -        - │
>   ├────────────────────────────────────────────────────────────────────────────────────────────────┤
>   │ ✔  template-tests/feedback.cy.js             73ms        1        1        -        -        - │
>   ├────────────────────────────────────────────────────────────────────────────────────────────────┤
>   │ ✔  template-tests/global-search.cy.js        57ms        1        1        -        -        - │
>   ├────────────────────────────────────────────────────────────────────────────────────────────────┤
>   │ ✔  template-tests/landing.cy.js             00:01        1        1        -        -        - │
>   ├────────────────────────────────────────────────────────────────────────────────────────────────┤
>   │ ✔  template-tests/vocab-home.cy.js          00:01        1        1        -        -        - │
>   ├────────────────────────────────────────────────────────────────────────────────────────────────┤
>   │ ✔  template-tests/vocab-search.cy.js         73ms        1        1        -        -        - │
>   └────────────────────────────────────────────────────────────────────────────────────────────────┘
>     ✖  6 of 16 failed (38%)                     00:15       16        7        6        3        -  
> ```
> 
> so five of the accessibility tests are now failing. I think this regression should be fixed before this can be merged.

1) I added the README.md file and included it along with the directory.

2) I adjusted the naming as per your suggestion, as the suggestion was well-founded. I also updated all the parts in the configurations that were affected by this change.

3) Since it does not cause regression and it would require further investigation, it is better to address this at a later time.

4) In the main branch, the skip functionality was intentionally enabled for most of the accessibility tests, meaning the tests were not meant to be run as there was nothing to test yet. However, I set up the feature branch to behave similarly.

Merge?

osma commented:
>     I added the README.md file and included it along with the directory.

Excellent! But the file is missing a final newline. Could you please add it?

>     I adjusted the naming as per your suggestion, as the suggestion was well-founded. I also updated all the parts in the configurations that were affected by this change.

:+1:

>     Since it does not cause regression and it would require further investigation, it is better to address this at a later time.

The code itself didn't change, but this is a regression in the sense that SonarCloud now complains about it when previously it didn't. I think it didn't look inside the `cypress` directory when it was at the root level, but now it does check `tests/cypress`.

Would this be simple to fix? Seems like a pretty trivial problem.

>     In the main branch, the skip functionality was intentionally enabled for most of the accessibility tests, meaning the tests were not meant to be run as there was nothing to test yet. However, I set up the feature branch to behave similarly.

:+1:

Now only the `template/error.cy.js` test is failing, which also happens on the `skosmos-3` branch, so there is no regression but that test should be fixed or skipped separate from this PR.
